src/app/encoder.cpp

```diff
@@ -0,0 +1,170 @@
+    {
+        ChipLogError(Zcl, "Could not encode buffer due to insufficient space");
+    }
+    else
+    {
+        ChipLogProgress(Zcl, "Successfully encoded APS frame of size %d bytes", buf.Written());
+    }
+
+    return (buf.Fit() || buf.Size() == 0) ? buf.Written() : 0; // 0 size buffer implies measure case.
+}
+
+uint16_t encodeApsFrame(uint8_t * buffer, uint16_t buf_length, EmberApsFrame * apsFrame)
+{
+    if (buffer == nullptr)
+    {
+        buf_length = 0; // This is the measuring case.
+    }
+    BufBound buf = BufBound(buffer, buf_length);
+    return doEncodeApsFrame(buf, apsFrame->profileId, apsFrame->clusterId, apsFrame->sourceEndpoint, apsFrame->destinationEndpoint,
+                            apsFrame->options, apsFrame->groupId, apsFrame->sequence, apsFrame->radius);
+}
+
+uint32_t _encodeOnOffCommand(uint8_t * buffer, uint32_t buf_length, uint8_t command, uint8_t destination_endpoint)
+{
+    uint32_t result = 0;
+    // pick cluster id as 6 for now.
+    // pick source and destination end points as 1 for now.
+    // Profile is 65535 because that matches our simple generated code, but we
+    // should sort out the profile situation.
+
+    if (buffer == nullptr)
+    {
+        buf_length = 0; // This is the measuring case.
+    }
+
+    BufBound buf = BufBound(buffer, buf_length);
+    result       = doEncodeApsFrame(buf, 65535, 6, 1, destination_endpoint, 0, 0, 0, 0);
+    if (result == 0)
+    {
+        ChipLogError(Zcl, "Error encoding aps frame result %" PRIu32 "\n", result);
+        result = 0;
+        return result;
+    }
+
+    // This is a cluster-specific command so low two bits are 0b01.  The command
+    // is standard, so does not need a manufacturer code, and we're sending
+    // client to server, so all the remaining bits are 0.
+    buf.Put(0x1);
+
+    // Transaction sequence number.  Just pick something.
+    buf.Put(0x1);
+
+    buf.Put(&command, sizeof(uint8_t));
+
+    result = buf.Fit() ? buf.Written() : 0;
+    if (result == 0)
+    {
+        ChipLogError(Zcl, "Error encoding on / off cmd");
+    }
+    return result;
+}
+
+uint32_t encodeOffCommand(uint8_t * buffer, uint32_t buf_length, uint8_t destination_endpoint)
+{
+    return _encodeOnOffCommand(buffer, buf_length, 0, destination_endpoint);
+};
+
+uint32_t encodeOnCommand(uint8_t * buffer, uint32_t buf_length, uint8_t destination_endpoint)
+{
+    return _encodeOnOffCommand(buffer, buf_length, 1, destination_endpoint);
+}
+
+uint32_t encodeToggleCommand(uint8_t * buffer, uint32_t buf_length, uint8_t destination_endpoint)
+{
+    return _encodeOnOffCommand(buffer, buf_length, 2, destination_endpoint);
+}
+
+uint16_t encodeReadAttributesCommand(uint8_t * buffer, uint16_t buf_length, uint8_t destination_endpoint, uint8_t cluster_id,
+                                     uint16_t * attr_ids, uint16_t attr_id_count)
+{
+    if (buffer == nullptr)
+    {
+        buf_length = 0; // This is the measuring case.
```

Comment:
Again, there is no such case here, per API docs...  And if this is meant to handle that case, it needs various other changes too, not just this change.